plugins/field-colour/src/blocks/colourBlend.ts

```diff
@@ -0,0 +1,271 @@
+    const c2 = generator.valueToCode(block, 'COLOUR2', PHP.Order.NONE) || "'#000000'";
+    const ratio = generator.valueToCode(block, 'RATIO', PHP.Order.NONE) || 0.5;
+    const functionName = generator.provideFunction_(
+        'colour_blend',
+        `
+  function ${generator.FUNCTION_NAME_PLACEHOLDER_}($c1, $c2, $ratio) {
+    $ratio = max(min($ratio, 1), 0);
+    $r1 = hexdec(substr($c1, 1, 2));
+    $g1 = hexdec(substr($c1, 3, 2));
+    $b1 = hexdec(substr($c1, 5, 2));
+    $r2 = hexdec(substr($c2, 1, 2));
+    $g2 = hexdec(substr($c2, 3, 2));
+    $b2 = hexdec(substr($c2, 5, 2));
+    $r = round($r1 * (1 - $ratio) + $r2 * $ratio);
+    $g = round($g1 * (1 - $ratio) + $g2 * $ratio);
+    $b = round($b1 * (1 - $ratio) + $b2 * $ratio);
+    $hex = '#';
+    $hex .= str_pad(dechex($r), 2, '0', STR_PAD_LEFT);
+    $hex .= str_pad(dechex($g), 2, '0', STR_PAD_LEFT);
+    $hex .= str_pad(dechex($b), 2, '0', STR_PAD_LEFT);
+    return $hex;
+  }
+  `,
+    );
+    const code = functionName + '(' + c1 + ', ' + c2 + ', ' + ratio + ')';
+    return [code, PHP.Order.FUNCTION_CALL];
+}
+
+/**
+ * Python generator definition.
+ * @param block 
+ * @param generator 
+ * @returns 
+ */
+export function pythonGenerator(
+    block: Blockly.Block,
+    generator: Python.PythonGenerator,
+): [string, Python.Order] {
+    // Blend two colours together.
+    const functionName = generator.provideFunction_(
+        'colour_blend',
+        `
+  def ${generator.FUNCTION_NAME_PLACEHOLDER_}(colour1, colour2, ratio):
+    r1, r2 = int(colour1[1:3], 16), int(colour2[1:3], 16)
+    g1, g2 = int(colour1[3:5], 16), int(colour2[3:5], 16)
+    b1, b2 = int(colour1[5:7], 16), int(colour2[5:7], 16)
+    ratio = min(1, max(0, ratio))
+    r = round(r1 * (1 - ratio) + r2 * ratio)
+    g = round(g1 * (1 - ratio) + g2 * ratio)
+    b = round(b1 * (1 - ratio) + b2 * ratio)
+    return '#%02x%02x%02x' % (r, g, b)
+  `,
+    );
+    const colour1 =
+        generator.valueToCode(block, 'COLOUR1', Python.Order.NONE) || "'#000000'";
+    const colour2 =
+        generator.valueToCode(block, 'COLOUR2', Python.Order.NONE) || "'#000000'";
+    const ratio = generator.valueToCode(block, 'RATIO', Python.Order.NONE) || 0;
+    const code =
+        functionName + '(' + colour1 + ', ' + colour2 + ', ' + ratio + ')';
+    return [code, Python.Order.FUNCTION_CALL];
+}
+
+/**
+ * Install the `colour_blend` block and all of its dependencies.
+ */
+export function installBlock(generators: Generators = {}) {
+    registerColourField();
+    Blockly.common.defineBlocksWithJsonArray([jsonDef]);
+    installGenerators(generators, blockName,
+        jsGenerator,
+        dartGenerator,
+        luaGenerator,
+        phpGenerator,
+        pythonGenerator);
+    // TODO: Should the reserved words be passed in at the same time as the generator
+    // for the installGenerators function?
+    generators.dart?.addReservedWords('Math');
```

Comment:
I'd be tempted to put the reserved words in a property on the generator function and let `installGenerators` retrieve them from there, but I know that some of my colleagues have this funny idea that functions shouldn't be treated like normal objects…

I observe that doing that would make this install function less useful as a guide to how to install the required individual pieces—though in fact the `installGenerators` function already does that to large extent.

I think I might favour doing it inline:
```JS
  if(generators.dart) generators.dart.forBlock[blockName] = jsGenerator;
  if(generators.lua) generators.lua.forBlock[blockName] = luaGenerator;
  //..
```
(With judicious choice of variable names it may be possible to keep each of these on a single line.)

Reply:
If I do it inline for `colour_picker`, it looks like this:
```

export function installBlock(generators: Generators = {}) {
    registerColourField();
    Blockly.common.defineBlocks([blockDef]);
    if (generators.javascript) generators.javascript.forBlock[BLOCK_NAME] = jsGenerator;
    if (generators.dart) generators.dart.forBlock[BLOCK_NAME] = dartGenerator;
    if (generators.lua) generators.lua.forBlock[BLOCK_NAME] = luaGenerator;
    if (generators.php) generators.php.forBlock[BLOCK_NAME] = phpGenerator;
    if (generators.python) generators.python.forBlock[BLOCK_NAME] = pythonGenerator;
}
```

And if I extend that to a case where I *do* have reserved words, like `colour_blend`, I get this:
```
export function installBlock(generators: Generators = {}) {
    registerColourField();
    Blockly.common.defineBlocks([blockDef]);
    if (generators.javascript) generators.javascript.forBlock[BLOCK_NAME] = jsGenerator;
    if (generators.dart) {
        generators.dart.forBlock[BLOCK_NAME] = dartGenerator;
        generators.dart.addReservedWords('Math');
    }
    if (generators.lua) generators.lua.forBlock[BLOCK_NAME] = luaGenerator;
    if (generators.php) generators.php.forBlock[BLOCK_NAME] = phpGenerator;
    if (generators.python) generators.python.forBlock[BLOCK_NAME] = pythonGenerator;
}
```

It's verbose, but has several advantages:
- Keeps reserved words and block generator installation together.
- The `installBlock` function actually shows everything it's doing
  -  As you observed, `installGenerators` was obfuscating.

The disadvantages are:
- Lots of text to copy.
- Lots of duplicated code.

I think this is worth doing, even if it adds duplicated code. The types did help me avoid installing a dart block generator into python, and that makes me happy. I think it's worth the change. @maribethb do you have any thoughts or objections?